pkgs/development/interpreters/python/fetch-python-requirements.nix

```diff
@@ -0,0 +1,181 @@
+# fetchPythonRequirements downlaods python packages specified by a list of
+# pip-style python requirements
+# It also requires a maximum date 'maxDate' being specified.
+# The result will be as if `pip download` would have been executed
+# at the point in time specified by maxDate.
+# This is ensured by putting pip behind a local proxy filtering the
+# api responses from pypi.org to only contain files for which the
+# release date is lower than the specified maxDate.
+
+{
+  # this specifies the python version for which the packages should be downloaded
+  python,
+}:
+
+{
+  lib
+, cacert
+, curl
+, stdenv
+, buildPackages
+, python3
+, ...
+}:
+let
+
+  # we use mitmproxy to filter the pypi responses
+  pythonWithMitmproxy = python3.withPackages (ps: [ ps.mitmproxy ps.python-dateutil ]);
+
+  fetchPythonRequirements = {
+    maxDate,        # maximum release date for packages
+    outputHash,     # hash for the fixed output derivation
+    requirements,   # list of strings of specs
+
+    # restrict to binary releases (.whl)
+    # this allows buildPlatform independent fetching
+    onlyBinary ? false,
+
+    # additional flags for `pip download`.
+    # for reference see: https://pip.pypa.io/en/stable/cli/pip_download/
+    pipFlags ? [],
+
+    # Throw an error by default to warn the user about the correct usage
+    targetSystem ? throw ''
+      'targetSystem' must be specified for fetchPythonRequirements.
+      Pass only a constant string like for example "x86_64-linux".
+      Do not pass `stdenv.targetPlatform.system` or `builtins.currentSystem`.
+      This prevents the fetcher from being executed on the wrong platform.
+    '',
+  }:
+
+    # specifying `--platform` for pip download is only allowed in combination with `--only-binary :all:`
+    # therefore, if onlyBinary is disabled, we must enforce targetPlatform == buildPlatform to ensure reproducibility
+    if ! onlyBinary && targetSystem != stdenv.buildPlatform.system then
+      throw ''
+        fetchPythonRequirements cannot fetch sdist packages for ${targetSystem} on a ${stdenv.buildPlatform.system}.
+        Either build on a ${targetSystem} or set `onlyBinary = true`.
+      ''
+    else
+    let
+      # map nixos system strings to python platforms
+      sysToPlatforms = {
+        "x86_64-linux" = [
+          "manylinux1_x86_64"
+          "manylinux2010_x86_64"
+          "manylinux2014_x86_64"
+        ];
+        "x86_64-darwin" =
+          lib.forEach (lib.range 0 15) (minor: "macosx_10_${builtins.toString minor}_x86_64");
+        "aarch64-linux" = [
+          "manylinux1_aarch64"
+          "manylinux2010_aarch64"
+          "manylinux2014_aarch64"
+        ];
+      };
+      platforms = if sysToPlatforms ? "${targetSystem}" then sysToPlatforms."${targetSystem}" else throw ''
```

Comment:
Does this work for non manylinux wheels? The only system I can think of right now that uses this is alpine and I am not sure if nixos static is that or still manylinux.

Reply:
Oh, I haven't even known about linux wheels until you mentioned it. Added them via 5893d09cccbddc8c44848750554595c64e73082a

Reply:
Wait a sec, do we even support `linux` wheels in nixpkgs? If not, it should probably not be listed here.